Framework/Muon/src/MuonAlgorithmHelper.cpp

```diff
@@ -0,0 +1,241 @@
+
+using namespace Mantid::Kernel;
+using namespace Mantid::API;
+using Mantid::Types::Core::DateAndTime;
+
+/**
+ * Return a first period MatrixWorkspace in a run workspace. If the run
+ * workspace has one period
+ * only - it is returned.
+ * @param ws :: Run workspace
+ */
+MatrixWorkspace_sptr firstPeriod(Workspace_sptr ws) {
+  if (auto group = boost::dynamic_pointer_cast<WorkspaceGroup>(ws)) {
+    return boost::dynamic_pointer_cast<MatrixWorkspace>(group->getItem(0));
+  } else {
+    return boost::dynamic_pointer_cast<MatrixWorkspace>(ws);
+  }
+}
+
+std::string getRunLabel(const Mantid::API::Workspace_sptr &ws) {
+  const std::vector<Mantid::API::Workspace_sptr> wsList{ws};
+  return getRunLabel(wsList);
+}
+
+/**
+ * Get a run label for a list of workspaces.
+ * E.g. for MUSR data of runs 15189, 15190, 15191 it will look like
+ * MUSR00015189-91.
+ * (Assumes all runs have the same instrument)
+ * @param wsList :: [input] Vector of workspace pointers
+ * @return :: run label
+ * @throws std::invalid_argument if empty list given
+ */
+std::string getRunLabel(const std::vector<Workspace_sptr> &wsList) {
+  if (wsList.empty())
+    throw std::invalid_argument("Unable to run on an empty list");
+
+  const std::string instrument =
+      firstPeriod(wsList.front())->getInstrument()->getName();
+
+  // Extract the run numbers
+  std::vector<int> runNumbers;
+  int numWorkspaces = static_cast<int>(wsList.size());
+  for (int i = 0; i < numWorkspaces; i++) {
+    int runNumber = firstPeriod(wsList[i])->getRunNumber();
+    runNumbers.push_back(runNumber);
+  }
+
+  return getRunLabel(instrument, runNumbers);
+}
+
+/**
+ * Get a run label for a given instrument and list of runs.
+ * E.g. for MUSR data of runs 15189, 15190, 15191 it will look like
+ * MUSR00015189-91.
+ * (Assumes all runs have the same instrument)
+ * @param instrument :: [input] instrument name
+ * @param runNumbers :: [input] List of run numbers
+ * @return :: run label
+ * @throws std::invalid_argument if empty run list given
+ */
+std::string getRunLabel(const std::string &instrument,
+                        const std::vector<int> &runNumbers) {
+  if (runNumbers.empty()) {
+    throw std::invalid_argument("Cannot run on an empty list");
+  }
+
+  // Find ranges of consecutive runs
+  auto ranges = findConsecutiveRuns(runNumbers);
+
+  // Zero-padding for the first run
+  int zeroPadding;
+  try {
+    zeroPadding = ConfigService::Instance()
+                      .getInstrument(instrument)
+                      .zeroPadding(ranges.begin()->first);
+  } catch (const Mantid::Kernel::Exception::NotFoundError &) {
+    // Old muon instrument without an IDF - default to 3 zeros
+    zeroPadding = 3;
+  }
+
+  // Begin string output with full label of first run
+  std::ostringstream label;
+  label << instrument;
+  label << std::setw(zeroPadding) << std::setfill('0') << std::right;
+
+  for (auto range : ranges) {
+    std::string firstRun = std::to_string(range.first);
```

Comment:
This might make sense as its own function, `std::string createStringFromRange(const std::pair<int, int> &range)`.
This would help simplify this entire code to something similar to:
```
std::ostream_iterator<std::string> addRunIterator(label,", ");
auto runStartIterator = boost::make_transform_iterator(ranges.begin(), createStringFromRange);
auto runEndIterator = boost::make_transform_iterator(ranges.end(), createStringFromRange);
std::copy(runStartIterator, runEndIterator, addRunIterator);
```

Note this specific idea would require including <boost/iterator/transform_iterator.hpp>